.github/workflows/CI.yml

```diff
@@ -45,29 +47,57 @@ jobs:
-          path: ~/.cache/bazel
+          key: build-cache.${{ matrix.os }}.${{ matrix.arch }}.julia-${{ matrix.version }}.ray-${{ steps.ray-commit.outputs.head_sha }}.hash-${{ hashFiles('deps/WORKSPACE.bazel.tpl', 'deps/BUILD.bazel') }}
+          path: ~/.cache
           restore-keys: |
-            bazel-cache.${{ matrix.os }}.${{ matrix.arch }}.julia-${{ matrix.version }}.ray-${{ steps.ray-commit.outputs.head_sha }}.
-            bazel-cache.${{ matrix.os }}.${{ matrix.arch }}.julia-${{ matrix.version }}.
-            bazel-cache.${{ matrix.os }}.${{ matrix.arch }}.
+            build-cache.${{ matrix.os }}.${{ matrix.arch }}.julia-${{ matrix.version }}.ray-${{ steps.ray-commit.outputs.head_sha }}.
+            build-cache.${{ matrix.os }}.${{ matrix.arch }}.julia-${{ matrix.version }}.
+
+      # Based upon:
+      # https://docs.ray.io/en/releases-2.5.1/ray-contribute/development.html#building-ray-on-linux-macos-full
+      - name: Build Ray CLI
+        run: |
+          pip install --upgrade pip wheel
+          pushd deps/ray
+
+          # The Ray BUILD.bazel includes a bunch of `copy_to_workspace` rules which copy build output
+          # into the Ray worktree. When we only restore the Bazel cache then re-building causes these
+          # rules to be skipped resulting in `error: [Errno 2] No such file or directory`. By manually
+          # saving/restoring these files we can work around this.
+          RAY_GEN_CACHE_DIR=~/.cache/ray-generated
+          if [ -d "$RAY_GEN_CACHE_DIR" ]; then
+              dest=$(pwd)
+              pushd $RAY_GEN_CACHE_DIR
+              cp -rp --parents \
+                  python/ray/_raylet.so \
+                  python/ray/core/generated \
+                  python/ray/serve/generated \
+                  python/ray/core/src/ray/raylet/raylet \
+                  python/ray/core/src/ray/gcs \
+                  $dest
+              popd
+          fi
+
+          pushd python
+          pip install . --verbose  # Fresh build takes ~50 minutes on basic GH runner
+          popd
+
+          # By copying the entire Ray worktree we can easily restore missing files without having to
+          # delete the cache and build from scratch.
+          mkdir -p "$RAY_GEN_CACHE_DIR"
+          cp -rfp . "$RAY_GEN_CACHE_DIR"
 
-      # TODO: Takes an hour to build with the existing cache setup
-      # - name: Build Ray CLI
-      #   run: |
-      #     cd deps/ray/python
-      #     pip install -e . --verbose
-      - name: Install Ray CLI
-        run: pip install -U "ray[default]==2.5.1" "pydantic<2"
       - uses: julia-actions/julia-buildpkg@v1
       - uses: julia-actions/julia-runtest@v1
 
       # https://github.com/actions/cache/tree/main/save#always-save-cache
-      - uses: actions/cache/save@v3
-        if: always()
+      - name: Save build cache
+        uses: actions/cache/save@v3
+        if: always() && steps.build-cache.outputs.cache-hit != 'true'
```

Comment:
what does the second condition check? something about the `build-cache` step not succeeding?